mteb/encoder_interface.py

```diff
@@ -0,0 +1,60 @@
+from __future__ import annotations
+
+from typing import Any, Protocol, runtime_checkable
+
+import numpy as np
+import torch
+
+
+@runtime_checkable
+class Encoder(Protocol):
+    """The interface for an encoder in MTEB."""
+
+    def encode(
+        self, sentences: list[str], prompt: str, **kwargs: Any
+    ) -> torch.Tensor | np.ndarray:
+        """Encodes the given sentences using the encoder.
+
+        Args:
+            sentences: The sentences to encode.
+            prompt: The prompt to use. Useful for prompt-based models.
+            **kwargs: Additional arguments to pass to the encoder.
+
+        Returns:
+            The encoded sentences.
+        """
+        ...
+
+
+class EncoderWithQueryCorpusEncode(Encoder, Protocol):
+    """The interface for an encoder that supports encoding a queries and a corpus."""
+
+    def encode_queries(
+        self, queries: list[str], prompt: str, **kwargs: Any
+    ) -> torch.Tensor | np.ndarray:
+        """Encodes the given queries using the encoder.
+
+        Args:
+            queries: The queries to encode.
+            prompt: The prompt to use. Useful for prompt-based models.
+            **kwargs: Additional arguments to pass to the encoder.
+
+        Returns:
+            The encoded queries.
```

Comment:
I believe this atm. can return `list[np.ndarray] | np.ndarray | list[torch.Tensor] | torch.Tensor`, seems like we should simplify it a bit.

Reply:
I was confused by this when I started using mteb, but I missed any historical reasons for allowing multiple types.  They do mostly behave the same for iterating though

Reply:
Yeah sometimes the corpora contain a title and a text in a dictionary i.e. sth like `{'title': '...', 'text': '...' ... }`

Reply:
@Muennighoff seems like you commented on the wrong thread :)

Reply:
Actually, it seems like SentenceTransformer encode has a "prompt_name" as input (thanks to @isaac-chung for pointing that out). This can e.g. be:

```
my_model = SentenceTransformer("e5_model", prompts={“query”: “query: “, “passage”: “passage: “})

my_model.encode(sentences, prompt_name="query")
```

It seems like this can completely replace the `encode_corpus/encode_query` interface.

Reply:
Actually, this gives us an issue if we e.g. both want to customize the prompt according to "query" as well as "task". A solution might be to simply do:

```
my_model.encode(sentences, prompt_name="task_name,query")
```

Reply:
Same comment as above, not sure this lets us do a unique prompt *per-query* only *per task*